src/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.OSX.cs

```diff
@@ -174,16 +174,105 @@ private sealed class RunningInstance
+                    Interop.EventStream.FSEventStreamScheduleWithRunLoop(eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+                    Interlocked.Increment(ref scheduledStreamsCount);
+                }
+
+                public static void UnscheduleFromRunLoop(SafeEventStreamHandle eventStream)
+                {
+                    if (_watcherRunLoop != IntPtr.Zero)
+                    {
+                        // Always unschedule the RunLoop before cleaning up
+                        Interop.EventStream.FSEventStreamUnscheduleFromRunLoop(eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+                        Interlocked.Decrement(ref scheduledStreamsCount);
+
+                        if (scheduledStreamsCount == 0)
+                        {
+                            // Stop the FS event message pump
+                            Interop.RunLoop.CFRunLoopStop(_watcherRunLoop);
+                        }
+                    }
+                }
+
+                private static void WatchForFileSystemEventsThreadStart(object args)
                 {
-                    if (!_stopping && _watcherRunLoop != IntPtr.Zero)
+                    var inputArgs = (object[])args;
+                    var runLoopStarted = (ManualResetEventSlim)inputArgs[0];
+                    var _eventStream = (SafeEventStreamHandle)inputArgs[1];
+                    // Get this thread's RunLoop
+                    _watcherRunLoop = Interop.RunLoop.CFRunLoopGetCurrent();
+                    Debug.Assert(_watcherRunLoop != IntPtr.Zero);
+
+                    // Retain the RunLoop so that it doesn't get moved or cleaned up before we're done with it.
+                    IntPtr retainResult = Interop.CoreFoundation.CFRetain(_watcherRunLoop);
+                    Debug.Assert(retainResult == _watcherRunLoop, "CFRetain is supposed to return the input value");
+
+
+                    // Schedule the EventStream to run on the thread's RunLoop
+                    Interop.EventStream.FSEventStreamScheduleWithRunLoop(_eventStream, _watcherRunLoop, Interop.RunLoop.kCFRunLoopDefaultMode);
+
+                    runLoopStarted.Set();
+                    try
+                    {
+                        // Start the OS X RunLoop (a blocking call) that will pump file system changes into the callback function
+                        Interop.RunLoop.CFRunLoopRun();
+                    }
+                    finally
+                    {
+                        // Release the WatcherLoop Core Foundation object.
+                        lock (lockObject)
+                        {
+                            Interop.CoreFoundation.CFRelease(_watcherRunLoop);
+                            _watcherRunLoop = IntPtr.Zero;
+                        }
+                    }
+                }
+            }
+
+            private void CancellationCallback()
+            {
+                lock (StopLock) {
```

Comment:
Shouldn't this be the same lock object that's used to coordinate other activities within StaticWatcherRunLoopManager?